### PR TITLE
Support "any" step in NumberControl and RangeControl

### LINKED ISSUE #34816: Components: support `step="any"` in `NumberControl` and `RangeControl`
## What

Accept `any` as a valid value for the `step` prop of the `NumberControl` and `RangeControl` components

## Why

The `input[type=number]` and `input[type=range]` HTML elements support `any` as a value for the `step` attribute ([#](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/range#attr-step), [#](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/number#attr-step)).

`NumberControl` and `RangeControl` aim at being a superset of these components, and should therefore support the `step="any"` case. Currently, this is not supported and causes the value of the inputs to become `NaN`

Also, partially related: #33156

## A/C

- [x] Introduce support for `step="any"` to `NumberControl` and `RangeControl`, following the expected behaviour from the HTML spec
- [ ] Add Storybook examples for both components
- [x] Update the `step` prop docs in both components README:
	- [x] The property can now be of type `number | "any"`
	- [x] Add a paragraph explaining the how the component should behave when passing "any"
- [ ] Add a CHANGELOG entry in the `@wordpress/components` package

### PR DESCRIPTION
**Pull request by stokesman**
Closes #34816 

Currently, `NumberControl` allows `step="any"` but attempting to use the arrow keys or dragging to change the value results in `NaN`. This PR fixes that.

Support in `RangeControl` presents a bit of a design challenge https://github.com/WordPress/gutenberg/pull/34542#issuecomment-921983454 and alternate ideas on resolving it are welcome.

## How has this been tested?
Unit tests and manually comparing behavior of `NumberControl` with behavior of an `input[type="number]` element using the same `step` values.

### `NumberControl` support of "any" step with arrow keys and dragging
number-control-any-step-and-shift-rounding

### `RangeControl` switching between `step` of 1 and "any"

https://user-images.githubusercontent.com/9000376/134936903-30f3b1b0-1538-4878-bd9c-152493b7ed42.mp4

### `RangeControl` marks with "any" step
The mark handling code needed a tiny revision to treat `step="any"` as if the step was 1. While testing that I noticed that the styling for marks has regressed but it's a separate issue (#35153).

https://user-images.githubusercontent.com/9000376/134937266-1f8378af-495c-4e2b-92d7-122b1226ed41.mp4

## Types of changes
enhancement

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/components/src/number-control/index.js
Comment (stokesman) [inline packages/components/src/number-control/index.js]: These changes were mostly in interest of formatting for a slight LOC reduction. The destructuring didn't seem helpful and by using `delta` (instead of the verbose `directionBaseValue`) for the value from the axis of interest, it let the new line 133 to fit on a single line.

File: packages/components/src/number-control/index.js
Comment (stokesman) [inline packages/components/src/number-control/index.js:139]: The value of `delta` comes from react-use-gesture and is a decimal value. It's converted to an integer in order to preserve the `NumberControl`'s decimal value as entered (when `step="any"`). Done via `Math.ceil` etc. because single pixel movements produce values like: 
image

**Comment by ciampo** [comment]
Hey @stokesman , thank you for your PR!

> Currently, `NumberControl` allows `step="any"`

Would you be able to point me to where this is stated? In the README, the [`step` prop is listed as of type `Number`](https://github.com/WordPress/gutenberg/blob/23e3fc6500a48ea59294c04a4bb41eca9dcbe498/packages/components/src/number-control/README.md#step), and therefore it doesn't support `any` as a value.

> Along with that, it revises the rounding used while holding the `Shift` key

While I understand that this may be a more predictable behaviour for you, I'll have to look more into it and consider if this is a change that we'd want to make to the component. It may be confusing to users that got used to the current rounding logic and therefore be seen as a sort of breaking change.

(cc @gziolo @diegohaz  )

**Comment by stokesman** [comment]
Thanks for taking a look and for the feedback @ciampo.

>In the README, step is listed as of type Number, and therefore it doesn't support any as a value.

I wrote "allows `step="any"`" because it seems incidental rather than intentional. However, it also seems like a useful thing to support. The README also states that "NumberControl is an enhanced HTML `input[type="number]` element". Was lack of support for [`step="any"`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/number#attr-step) intended as an enhancement then?

> It may be confusing to users that got used to the current rounding logic and therefore be seen as a sort of breaking change.

I believe that most users of the feature only even know about it because they use the same feature in other applications. All other applications I know with the feature, do not round by an increased amount while holding <kbd>Shift</kbd>. So for those relatively advanced users, I think we'd avoid frustration with the unique implementation we currently have.

**Review by aaronrobertshaw — COMMENTED** [review]
@stokesman Thanks for working on this! 👍 

The updated NumberControl works as advertised. 

Personally, I like the consistent and predictable application of step values. That said, I can appreciate that this is a change away from the current behaviour that some users may not perceive as a bug and have grown accustomed to.

The docs for the [`shiftStep` prop](https://github.com/WordPress/gutenberg/blob/fix/number-control-any-step-and-shift-rounding/packages/components/src/number-control/README.md#shiftstep) don't seem to communicate any rounding of the value based off of it. So perhaps that aspect of the changes in this PR are more in line with the advertised functionality?

> Amount to increment by when the SHIFT key is held down. This shift value is a multiplier to the step value. For example, if the step value is 5, and shiftStep is 10, each jump would increment/decrement by 50.

Regarding the support of "any" I'll happily defer to @ciampo on that. 🙂

File: packages/components/src/number-control/index.js
Comment (aaronrobertshaw) [reply packages/components/src/number-control/index.js]: The destructuring of the delta does make it immediately clear which value, `x` or `y`, is being used. Could the line length issue or naming be improved without removing that clarity?

**Review by ciampo — COMMENTED** [review]
### About supporting `step="any"`

> The README also states that "NumberControl is an enhanced HTML input[type="number] element". Was lack of support for step="any" intended as an enhancement then?

Seen in this light, it does look like adding support for `any` as a valid value for the `step` prop would increase this component's adherence to the HTML `input` spec.

It would be good, then, to update the README:

- the type of the `step` prop should become something like `number | "any"`
- there should be a short paragraph explaining what to expect when setting `step="any"`

It'd be also good to add support for `any` to the Storybook example

### About the shift step rounding changes

@aaronrobertshaw made a very good point:

> The docs for the shiftStep prop don't seem to communicate any rounding of the value based off of it. So perhaps that aspect of the changes in this PR are more in line with the advertised functionality?

So I'd say that we can go ahead and implement this change — but the approach I would take would be slightly different: I would instead look at changing the logic directly in [the `useJumpStep` function](https://github.com/WordPress/gutenberg/blob/a7e2895829c16ecd77a5ba22d84f1dee1cfb0977/packages/components/src/utils/hooks/use-jump-step.js).

### Splitting across two separate PRs

Since the `step="any"` attribute [is also supported by `<input type="range" />`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/range#attr-step), similar changes to the ones applied to `NumberControl` in this PR should be also applied to `RangeControl`.

Also, since `useJumpStep` is also used by `RangeControl`, I would propose that:

- We keep this PR focused on adding support of the `step="any"` prop to both `NumberControl` and `RangeControl`
- We make changes to the shift step rounding logic in a separate PR, by changing the `useJumpStep` function

**Comment by stokesman** [comment]
Thank you both @aaronrobertshaw and @ciampo 🙇 

>- We keep this PR focused on adding support of the `step="any"` prop to both `NumberControl` and `RangeControl`

I like that idea and have pushed some changes in that direction. I've yet to add any changes for `RangeControl` and I'll draft this again for now.

>- We make changes to the shift step rounding logic in a separate PR, by changing the `useJumpStep` function

I'm all for the separate PR and sharing any appropriate abstractions but, if we change `useJumpStep`, I have a feeling it will be a nigh 100% rewrite. We can get into the details as progress warrants. In related news, I've just opened a [PR that incidentally removes the hook from `RangeControl`](34719) as a solution to [this recent issue](34363)

File: packages/components/src/number-control/index.js
Comment (stokesman) [reply packages/components/src/number-control/index.js]: Thank you! I brought back the destructure to `x` and `y`.

**Comment by ciampo** [comment]
> > We keep this PR focused on adding support of the step="any" prop to both NumberControl and RangeControl
>
> I like that idea and have pushed some changes in that direction. I've yet to add any changes for `RangeControl` and I'll draft this again for now.

Sounds good. I've opened #34816 to track this issue and add detailed acceptance criteria for this PR — let me know it's ready for another round of review!

> 
> > * We make changes to the shift step rounding logic in a separate PR, by changing the `useJumpStep` function
> 
> I'm all for the separate PR and sharing any appropriate abstractions but, if we change `useJumpStep`, I have a feeling it will be a nigh 100% rewrite. We can get into the details as progress warrants. In related news, I've just opened a [PR that incidentally removes the hook from `RangeControl`](34719) as a solution to [this recent issue](34363)

I'm going to have a look at #34719, thank you for the ping there!

I've opened #34817 to track the update of the `shift` step rounding logic separately. We can discuss the next steps there, and keep the focus in this PR on the `step="any"` support

**Comment by stokesman** [comment]
Not much seems to be needed to have `RangeControl` support `step="any"` but it introduces design issues since input tends to make great decimals. The first solution that came to mind would be to override the default of `showTooltip` and perhaps `withInputField` to `false` when `step="any"`. That would seem acceptable since the native input range is without either feature and using `step="any"` should mean the use case is one in which a particular value isn't crucial.

`step="any"` with tooltip and input | `step="any"` as proposed
----|----
image | image

I've pushed according changes in case anyone cares to try it but the screenshots should reveal all there is to know. If the approach seems good I'll still need to make docs/changelog edits before this is review-ready. Also I think it's better to wait on the outcome of #34719 before code review here as changes will be influenced.

File: packages/components/src/range-control/index.js
Comment (aaronrobertshaw) [inline packages/components/src/range-control/index.js]: Nitpick: I think this would be better split into two lines. Not sure it's official in the code style guidelines but it would make this consistent with the rest of the codebase.

```suggestion
		showTooltipProp = false;
		withInputField = false;
```

**Review by aaronrobertshaw — COMMENTED** [review]
Thanks for all the continued efforts on this PR @stokesman 👍 

Does the proposed removal of the tooltip and input field lead to accessibility concerns? I'd be interested to hear what others think on this front.

Also, it might be good to update the `RangeControl` storybook example with regards to `step` allowing the `any` value as per the change made for the `NumberControl`.

I appreciate you mentioned this wasn't quite ready for review yet. Hope you don't mind me taking another look while I had the chance.

File: packages/components/src/range-control/index.js
Comment (stokesman) [reply packages/components/src/range-control/index.js]: Thanks for the note on this. I'd been curious to see if it lints but might have forgotten about it.

**Comment by stokesman** [comment]
I've pushed some updates to move forward with the RangeControl support as proposed in my previous comment. I've updated the description of this PR with some new screen captures and a little detail about the RangeControl updates.

This potentially completes the PR aside from adding a changelog entry to note the enhancement. I went back over the A/C checklist in the issue @ciampo created and checked things off. I wasn't sure if we wanted to add a new separate story for `step="any"` in `NumberControl` or if the update to the existing story is sufficient. 

> Does the proposed removal of the tooltip and input field lead to accessibility concerns?

My thinking is that it does not, or that if it does, it's already possible to configure the component that way so this isn't a new thing. I'd be glad to hear what others think as well. I've certainly considered other ideas for the tooltip like truncation or rounding to lesser precision but those, as I see it, would compromise poorly and add complexity.

File: packages/components/src/number-control/README.md
Comment (ciampo) [inline packages/components/src/number-control/README.md]:
```suggestion
-   Default: `Infinity`
```

File: packages/components/src/number-control/README.md
Comment (ciampo) [inline packages/components/src/number-control/README.md]:
```suggestion
-   Default: `-Infinity`
```

File: packages/components/src/number-control/index.js
Comment (ciampo) [inline packages/components/src/number-control/index.js]: `baseStep` doesn't exist in `useJumpStep`'s signature — did you mean 

```suggestion
	const jumpStep = useJumpStep( {
		step: baseStep,
		shiftStep,
		isShiftStepEnabled,
	} );
```

File: packages/components/src/number-control/index.js
Comment (ciampo) [inline packages/components/src/number-control/index.js:99]: Minor nit: I think `undefined` is a better fit, as it would equal to not passing the second argument at all

```suggestion
				enableShift ? incrementalValue : undefined
```

File: packages/components/src/number-control/index.js
Comment (ciampo) [inline packages/components/src/number-control/index.js:108]: Just flagging that using `isShiftStepEnabled` is a change from the different behaviour, but I'd say a welcome one, since previously `isShiftStepEnabled` was being ignored when dragging the value. Nice find!

File: packages/components/src/number-control/index.js
Comment (ciampo) [inline packages/components/src/number-control/index.js:144]: Whatever we decide from the previous comment on passing `null` or `undefined` in the `constrainValue` function, should also be applied here

**Review by ciampo — COMMENTED** [review]
Thank you for keeping up the work on this PR!

I was testing the component on Storybook, and I noticed some unexpected behaviour:

https://user-images.githubusercontent.com/1083581/135124283-dd41356b-8527-4f54-b4ef-2cec8d498311.mp4

I left `step="1"` and set `shiftStep="2.5"`. The unexpected behaviours are:

- I started shift-stepping up until `12.5` — notice how the value, despite not being valid, doesn't get round-clamped to the closest valid value. After that, I pressed the arrow up (no shift), and the value went to `14` — I expected it to be `13.5`, given how it wasn't round-clamped before.
- Later on, the value was `20`. I shift-stepped up, the value went to `22.5`. I stepped up (no shift), and the value jumped to `24` (so far, same experience as the previous point). But then, I shift-stepped again, and the value jumped to `27.5`. I found this surprising, since from `24` I expected to jump to `26.5`.

In brief, I think that the round-clamping logic is not behaving as expected when using a mix of normal stepping and shift-stepping. Also, the same behaviour can be replicated with an integer as the shift step.

Witnessing this buggy behaviour makes me even more convinced that we need to write more tests for this component, and makes me doubt about keeping the shift-stepping functionality.

I also left a few inline comments.

I haven't gotten very far testing the `step=any` functionality and the `RangeControl` component in Storybook because of the unexpected round-clamping behaviour.

File: packages/components/src/number-control/index.js
Comment (stokesman) [reply packages/components/src/number-control/index.js]: 🤦 , Sorry you had to see that but thanks for catching it!

File: packages/components/src/number-control/index.js
Comment (stokesman) [reply packages/components/src/number-control/index.js:99]: It's all nullish to me and `constrainValue` 😄 . I would have committed this but the next comment about the same didn't also have the suggestion so I left them both out for now.

What I'd like to point out is that if we proceed with #34817 the `constrainValue` function will not even have a second parameter so this can be made moot. That's how it was in the original version of this PR and part of the reason I just wanted to do it all in one go.

**Comment by stokesman** [comment]
Thanks for the testing and feedback Marco! I've gone ahead with most of the suggestions you made. 

> In brief, I think that the round-clamping logic is not behaving as expected when using a mix of normal stepping and shift-stepping. Also, the same behaviour can be replicated with an integer as the shift step.

Unless it was partly due to the mistake I'd made (and you caught https://github.com/WordPress/gutenberg/pull/34542#discussion_r717668068) then I believe all that is existing and what this PR is helping us get closer to fixing. I have #34566 waiting in the wings that adds some tests and fixes for `roundClamp`. It was depending slightly on something from the original version of this PR (that also included the shift-step refinement) so I'd drafted it until those get worked out.

> …notice how the value, despite not being valid, doesn't get round-clamped to the closest valid value.

I think that is #33291 for which #34128 exists. (Also not sure if you expected the UA to do it. None that I know of do that for number inputs just range inputs (and that's only Firefox as far as I know)).

**Comment by stokesman** [comment]
I just went through and tested on the [github hosted storybook](https://wordpress.github.io/gutenberg/?path=/story/components-numbercontrol--default) and could reproduce exactly the results described in https://github.com/WordPress/gutenberg/pull/34542#pullrequestreview-765559978. Those are expected in that they are what we currently ship. We do have some unit tests that should fail given any shift-step changes (they did in [the original version of this PR](https://github.com/WordPress/gutenberg/compare/trunk...stokesman:fix/number-control-shift-step-rounding?expand=1#diff-e441a3604d6638af129de7d20f1a50980190dc1f463a95ff8f82438cde858da9) which included shift-step rounding changes).

File: packages/components/src/number-control/index.js
Comment (ciampo) [reply packages/components/src/number-control/index.js:99]: Let's leave it as-is, then. It was a minor nit anyway

**Review by ciampo — COMMENTED** [review]
> I just went through and tested on the github hosted storybook and could reproduce exactly the results described in #34542 (review). Those are expected in that they are what we currently ship

I can confirm that the behaviour that I flagged previously can be also replicated in production — so we shouldn't count it as a blocker for this PR.

> We do have some unit tests that should fail given any shift-step changes (they did in the original version of this PR which included shift-step rounding changes).

That's a good point (although they were not testing for the bug that I flagged earlier).

In that regard, thank you for flagging #34566 and #34128 — it'd be great if we could work on it as a follow-up, I believe those fixes and the additional tests are much needed, and I would argue they should take precedence over resuming the work in #34817.

---

I was ready to approve this PR, but I decided to give it a final look — and that's when I discovered another unexpected behaviour, which I believe should be fixed in this PR since it's related to `step="any"`.

Steps to reproduce:

- I set `step="any"`
- I set the value of the input to a non-integer number, e.4. `4.3`
- I increment/decrement the value via keyboard, works as expected
- I press `enter` to force the rounding — the `.3` part of the number is preserved, as expected
- I click-drag, the number increments/decrements as expected
- I click on the up/down arrows on the right side of the input — the value gets rounded to the closest integer. I think this behaviour is unexpected and should be updated to reflect the rest of the interactions listed above —  what do you think?


Here's a screencast of the behaviour that I just described

https://user-images.githubusercontent.com/1083581/135610079-49daf819-c717-411a-bb4b-48e9aeb0ab32.mp4

Finally, I think it would be nice if we could add tests specific to `steps="any"` (as part of this PR).

**Comment by stokesman** [comment]
Thanks for testing this once again @ciampo. 

> I click on the up/down arrows on the right side of the input — the value gets rounded to the closest integer. I think this behaviour is unexpected and should be updated to reflect the rest of the interactions listed above — what do you think?

I agree and it presents a bit of a dilemma. We can't (easily) have both shift-step working for clicks on those up/down arrows and have them behave as they should natively in regards to not rounding "any" steps. It's because the shift-step for those arrows happens the same way it did for `RangeControl`, by mutating the DOM step attribute and letting the browser handle it. To handle with our own logic is possible but not trivial.

I've pushed a change for what I think is the sensible way to proceed and that is removing the mutation of the step attribute. So clicking the arrows while holding shift won't shift-step. We could compromise and remove it only when `step="any"` but there doesn't seem to be much reason to bother. Given that the component is experimental and that it seems unlikely many folks expect to shift-step while clicking those arrows, I think it's fine to remove it. Additionally, mutating the attribute may not cause any known issues but it's a bit dirty as it can temporarily invalidate the input.

One more related note is it looks like we may be adding our own stepping arrows to `NumberControl` in which case having them shift-step would be relatively easy.
image

I've added some tests for `step="any"` with `NumberControl` to the PR. They are only for keyboard input because testing the drag or "spin" arrows in jsdom would be a fancy trick. I looked at adding some tests for `RangeControl` but can't see there's any worthwhile tests to make there.

**Review by ciampo — APPROVED** [review]
> Given that the component is experimental and that it seems unlikely many folks expect to shift-step while clicking those arrows, I think it's fine to remove it. Additionally, mutating the attribute may not cause any known issues but it's a bit dirty as it can temporarily invalidate the input.
> 
> One more related note is it looks like we may be adding our own stepping arrows to NumberControl in which case having them shift-step would be relatively easy.

Thank you for the detailed explanation. I agree with the whole reasoning here.

I can also confirm that the issue that I flagged in my previous comment is now fixed.

> I've added some tests for step="any" with NumberControl to the PR. They are only for keyboard input because testing the drag or "spin" arrows in jsdom would be a fancy trick. I looked at adding some tests for RangeControl but can't see there's any worthwhile tests to make there.

Thank you for this as well.

I am not 100% sure I like how we're hiding the number input in the `RangeControl` component when `step="any"`, but it's probably something we can iterate on in a separate PR as a follow-up.

Let's merge this 🚀  (and thank you again for the patience while working on this PR)

**Comment by stokesman** [comment]
Thank you @ciampo 🙇  I just pushed one more commit for a changelog entry (on a rebase). A test may have flaked but as soon as I can get it all green, I'll merge it.

One other thing, It looks like `useJumpStep` will become dead code when this merges. I suppose it won't hurt to leave it be for now but thought it worth noting.

**Comment by ciampo** [comment]
> One other thing, It looks like useJumpStep will become dead code when this merges. I suppose it won't hurt to leave it be for now but thought it worth noting.

Thank you for flagging it. Let's continue this conversation in #34817

**Pull request merged**
